cmd/lakefs/cmd/kv.go

```diff
@@ -0,0 +1,167 @@
+)
+
+var errInvalidParamValue = errors.New("invalid parameter value")
+
+var kvCmd = &cobra.Command{
+	Use:    "kv",
+	Short:  "Inspect lakeFS' Key-Value Store",
+	Hidden: true,
+}
+
+var kvGetCmd = &cobra.Command{
+	Use:   "get <partition> <path>",
+	Short: "Return the value for the given path under the given partition",
+	Args:  cobra.ExactArgs(GetCmdNumArgs),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		cfg := loadConfig()
+
+		pretty, err := cmd.Flags().GetBool("pretty")
+		if err != nil {
+			return err
+		}
+
+		ctx := cmd.Context()
+		kvParams, err := cfg.GetKVParams()
+		if err != nil {
+			return fmt.Errorf("KV params: %w", err)
+		}
+
+		kvStore, err := kv.Open(ctx, kvParams)
+		if err != nil {
+			return fmt.Errorf("failed to open KV store: %w", err)
+		}
+		defer kvStore.Close()
+
+		partitionKey := args[0]
+		path := args[1]
+		val, err := kvStore.Get(ctx, []byte(partitionKey), []byte(path))
+		if err != nil {
+			return fmt.Errorf("get failed: %w", err)
+		}
+		kvObj, err := kv.NewRecord(partitionKey, path, val.Value)
+		if err != nil {
+			return fmt.Errorf("KV record from value: %w", err)
+		}
+
+		encoder := json.NewEncoder(os.Stdout)
+		if pretty {
+			encoder.SetIndent("", "  ")
+		}
+		err = encoder.Encode(kvObj)
+		if err != nil {
+			return fmt.Errorf("json.Marshal failed: %w", err)
+		}
+
+		return nil
+	},
+}
+
+var kvScanCmd = &cobra.Command{
+	Use:   "scan <partition> [<path>]",
+	Short: "Scan through keys and values under the given partition. An optional path can be specified as a starting point (inclusive)",
+	Args:  cobra.RangeArgs(ScanCmdMinArgs, ScanCmdMaxArgs),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		cfg := loadConfig()
+
+		limit, err := cmd.Flags().GetInt("limit")
+		if err != nil {
+			return err
+		}
+		until, err := cmd.Flags().GetString("until")
+		if err != nil {
+			return err
+		}
+		pretty, err := cmd.Flags().GetBool("pretty")
+		if err != nil {
+			return err
+		}
+
+		partitionKey := args[0]
+		var start []byte
+		if len(args) > ScanCmdMinArgs {
+			start = []byte(args[1])
+		}
+
+		if len(until) > 0 && until < string(start) {
+			return fmt.Errorf("`until` cannot precede `path`: %w", errInvalidParamValue)
+		}
+
+		ctx := cmd.Context()
+		kvParams, err := cfg.GetKVParams()
+		if err != nil {
+			return fmt.Errorf("KV params: %w", err)
+		}
+
+		kvStore, err := kv.Open(ctx, kvParams)
+		if err != nil {
+			return fmt.Errorf("failed to open KV store: %w", err)
+		}
+		defer kvStore.Close()
+
+		iter, err := kvStore.Scan(ctx, []byte(partitionKey), start)
+		if err != nil {
+			return fmt.Errorf("scan failed: %w", err)
+		}
+		defer iter.Close()
+
+		num := 0
+		for iter.Next() {
+			entry := iter.Entry()
+			if len(until) > 0 && string(entry.Key) > until {
+				break
+			}
+			kvObj, err := kv.NewRecord(partitionKey, string(entry.Key), entry.Value)
+			if err != nil {
+				return fmt.Errorf("KV record from value: %w", err)
+			}
+			encoder := json.NewEncoder(os.Stdout)
```

Comment:
We should have a new encoder once outside the loop

Reply:
Correct - done